src/app/clusters/closure-dimension-server/closure-dimension-cluster-logic.cpp

```diff
@@ -19,93 +19,302 @@
+    bool validateIncomingCurrentStateSpeed    = false;
+
+    bool requirePositioningUpdate = false;
+    bool requireLatchUpdate       = false;
+    bool requireSpeedUpdate       = false;
+
+    /*
+        Determine checks that need to be executed based on current and provided values
+    */
 
-    if (currentState.position.HasValue())
+    // Determine CurrentState member values, if a value is present, if it is null or if it has a valid value, to determine what
+    // checks need to be executed.
+    if (!clusterCurrentState.IsNull())
     {
+
+        // If the CurrentState has a value, we need to check internal members to validate if we need to mark the attribute as dirty.
+        validateClusterCurrentStateMembers = true;
+
+        validateClusterCurrentStatePosition = clusterCurrentState.Value().position.HasValue();
+        validateClusterCurrentStateLatch    = clusterCurrentState.Value().latch.HasValue();
+        validateClusterCurrentStateSpeed    = clusterCurrentState.Value().speed.HasValue();
+    }
+
+    // Determine CurrentState member values, if a value is present, if it is null or if it has a valid value, to determine what
+    // checks need to be executed.
+    if (!incomingCurrentState.IsNull())
+    {
+        // If the provided CurrentState has a value, we need to check its members to validate
+        validateIncomingCurrentStateMembers = true;
+
+        validateIncomingCurrentStatePosition = incomingCurrentState.Value().position.HasValue();
+        validateIncomingCurrentStateLatch    = incomingCurrentState.Value().latch.HasValue();
+        validateIncomingCurrentStateSpeed    = incomingCurrentState.Value().speed.HasValue();
+    }
+
+    /*
+        Validate incoming values and FeatureMap Conformance
+    */
+
+    // Validate the incoming Position value has valid input parameters and FeatureMap conformance.
+    if (validateIncomingCurrentStatePosition)
+    {
+        //  If the position member is present in the incoming CurrentState, we need to check if the Positioning
+        //  feature is supported by the closure. If the Positioning feature is not supported, return an error.
         VerifyOrReturnError(mConformance.HasFeature(Feature::kPositioning), CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
-        VerifyOrReturnError(currentState.position.Value() <= PERCENT100THS_MAX_VALUE, CHIP_ERROR_INVALID_ARGUMENT);
+
+        // We don't need to check if values are present since the check was done above.
+        const Percent100ths & position = incomingCurrentState.Value().position.Value();
+        VerifyOrReturnError(position <= kPercents100thsMaxValue, CHIP_ERROR_INVALID_ARGUMENT);
+
+        // We need to update the cluster positioning value if
+        // - cluster does not have value; the value will necessarily be updated
+        // - cluster and incoming values are different
+        requirePositioningUpdate = !validateClusterCurrentStatePosition || clusterCurrentState.Value().position.Value() != position;
```

Comment:
Why do we care about validateClusterCurrentStatePosition?  Why can't we just do:
```suggestion
        requirePositioningUpdate = validateClusterCurrentStateMembers && clusterCurrentState.Value().position != position;
```

But also: if our current position has a value and the incoming position has no value, is it expected that the current position does not change and we don't mark things dirty... unless something else has changed and _then_ we will change the current position?  It's really not clear to me what behavior is being aimed for here.

It _does_ however, seem to me that the logic here is _way_ more complicated than it needs to be, given that we have a struct type that supports both `operator==` and `operator=`, right?  We should just be able to compute the new value of current position, mark dirty if that's not equal to the old value, assign to the old value....

Reply:
modified the logic please check